### DISCUSSION TITLE
Welcome to tespy Discussions!

### DISCUSSION THREAD

**Discussion by fwitte** [oemof/tespy#304] [Announcements]
## 👋 Welcome to the TESPy user forum
  We’re using Discussions as a place to connect with other users and developers of our community. Our goals are 

* to give you a place to ask questions you’re wondering about,
* to connect you with the community, so you can benefit from each other,
* to share knowledge and new ideas with the community,
* to provide a place for important updates and news from the devs.

To get started, simply open a new discussion or go through the list of topics and engage with others.

**Comment by Mamun5001** [comment]
#439 
Hi all users of Tespy
I am a beginner user of Tespy. I am trying to do one thing 'How can I find heat and efficiency by changing the overall temperature level in the heating system'. I have taken the DH network which is solved 
(https://github.com/oemof/oemof-examples/blob/master/oemof_examples/tespy/district_heating/dhs.py)

I have solved these 2 conditions 
1. changing ambient temperature 
2. heat load varies 

Please help me write the code, from that I can get my idea. Thank you

**Comment by Adefunmike** [comment]
Please i need help with the code below

from tespy.networks import Network
from tespy.components import (
    Sink, Source, Splitter, Compressor, Condenser, Pump, HeatExchangerSimple,
    Valve, Drum, HeatExchanger, CycleCloser
)
from tespy.connections import Connection, Ref
from tespy.tools.characteristics import CharLine
from tespy.tools.characteristics import load_default_char as ldc
from tespy.tools import document_model

import numpy as np
import pandas as pd

# %% network

# Define the working fluids, units, and mass flow rates
nw = Network(
    fluids=['water', 'NH3', 'air'], T_unit='C', p_unit='bar', h_unit='kJ / kg',
    m_unit='kg / s'
)

# %% components

# Define sources & sinks for the network
cc = CycleCloser('coolant cycle closer')
cc_cons = CycleCloser('consumer cycle closer')
amb = Source('ambient air')
amb_out1 = Sink('sink ambient 1')
amb_out2 = Sink('sink ambient 2')

# Components from your schematic
sp = Splitter('splitter')
fan = Compressor('fan')
cd = Condenser('condenser')
dhp = Pump('district heating pump')
cons = HeatExchangerSimple('consumer')
ves = Valve('valve')
dr = Drum('drum')
ev = HeatExchanger('evaporator')
su = HeatExchanger('superheater')
erp = Pump('evaporator reciculation pump')
cp1 = Compressor('compressor 1')
cp2 = Compressor('compressor 2')
ic = HeatExchanger('intercooler')

# %% connections

# Consumer system connections
c_in_cd = Connection(cc, 'out1', cd, 'in1')
cb_dhp = Connection(cc_cons, 'out1', dhp, 'in1')
dhp_cd = Connection(dhp, 'out1', cd, 'in2')
cd_cons = Connection(cd, 'out2', cons, 'in1')
cons_cf = Connection(cons, 'out1', cc_cons, 'in1')

nw.add_conns(c_in_cd, cb_dhp, dhp_cd, cd_cons, cons_cf)

# Condenser to evaporator connection
cd_ves = Connection(cd, 'out1', ves, 'in1')
nw.add_conns(cd_ves)

# Evaporator system connections
ves_dr = Connection(ves, 'out1', dr, 'in1')
dr_erp = Connection(dr, 'out1', erp, 'in1')
erp_ev = Connection(erp, 'out1', ev, 'in2')
ev_dr = Connection(ev, 'out2', dr, 'in2')
dr_su = Connection(dr, 'out2', su, 'in2')
nw.add_conns(ves_dr, dr_erp, erp_ev, ev_dr, dr_su)

amb_fan = Connection(amb, 'out1', fan, 'in1')
fan_sp = Connection(fan, 'out1', sp, 'in1')
sp_su = Connection(sp, 'out1', su, 'in1')
su_ev = Connection(su, 'out1', ev, 'in1')
ev_amb_out = Connection(ev, 'out1', amb_out1, 'in1')
nw.add_conns(amb_fan, fan_sp, sp_su, su_ev, ev_amb_out)

# Evaporator to compressor connections
su_cp1 = Connection(su, 'out2', cp1, 'in1')
nw.add_conns(su_cp1)

# Compressor-system connections
cp1_he = Connection(cp1, 'out1', ic, 'in1')
he_cp2 = Connection(ic, 'out1', cp2, 'in1')
cp2_c_out = Connection(cp2, 'out1', cc, 'in1')
sp_ic = Connection(sp, 'out2', ic, 'in2')
ic_out = Connection(ic, 'out2', amb_out2, 'in1')
nw.add_conns(cp1_he, he_cp2, sp_ic, ic_out, cp2_c_out)

# %% component parametrization

# Condenser
cd.set_attr(pr1=0.99, pr2=0.99, ttd_u=5, design=['pr2', 'ttd_u'],
            offdesign=['zeta2', 'kA_char'])
dhp.set_attr(eta_s=0.8, design=['eta_s'], offdesign=['eta_s_char'])
cons.set_attr(pr=0.99, design=['pr'], offdesign=['zeta'])

# Air fan
fan.set_attr(eta_s=0.65, design=['eta_s'], offdesign=['eta_s_char'])

# Evaporator system
kA_char1 = ldc('heat exchanger', 'kA_char1', 'DEFAULT', CharLine)
kA_char2 = ldc('heat exchanger', 'kA_char2', 'EVAPORATING FLUID', CharLine)
ev.set_attr(pr1=0.999, pr2=0.99, ttd_l=5,
            kA_char1=kA_char1, kA_char2=kA_char2,
            design=['pr1', 'ttd_l'], offdesign=['zeta1', 'kA_char'])
su.set_attr(pr1=0.999, pr2=0.99, ttd_u=2, design=['pr1', 'pr2', 'ttd_u'],
            offdesign=['zeta1', 'zeta2', 'kA_char'])
erp.set_attr(eta_s=0.8, design=['eta_s'], offdesign=['eta_s_char'])

# Compressor system
cp1.set_attr(eta_s=0.85, design=['eta_s'], offdesign=['eta_s_char'])
cp2.set_attr(eta_s=0.9, pr=3, design=['eta_s'], offdesign=['eta_s_char'])
ic.set_attr(pr1=0.99, pr2=0.999, design=['pr1', 'pr2'],
            offdesign=['zeta1', 'zeta2', 'kA_char'])

# %% connection parametrization

# Working fluid mass flow rates and pressures from your schematic
c_in_cd.set_attr(fluid={'air': 0, 'NH3': 1, 'water': 0})
cb_dhp.set_attr(T=60, p=10, fluid={'air': 0, 'NH3': 0, 'water': 1})
cd_cons.set_attr(T=90)

# Evaporator mass flow rate and pressure setup
erp_ev.set_attr(m=Ref(ves_dr, 1.25, 0), p0=5)
erp_ev.set_attr(m=Ref(ves_dr, 1.25, 0), p0=5)


su_cp1.set_attr(p0=5, state='g')

# Evaporator system hot side
amb_fan.set_attr(T=12, p=1, fluid={'air': 1, 'NH3': 0, 'water': 0},
                 offdesign=['v'])
sp_su.set_attr(offdesign=['v'])
ev_amb_out.set_attr(p=1, T=9, design=['T'])

# Compressor-system
he_cp2.set_attr(Td_bp=5, p0=20, design=['Td_bp'])
ic_out.set_attr(T=30, design=['T'])

# Set power based on your system schematic
cons.set_attr(Q=-200e3)

# %% Calculation

# Design mode solution
nw.solve('design')
nw.print_results()
nw.save('heat_pump_air')

# Off-design test
nw.solve('offdesign', design_path='heat_pump_air')
document_model(nw, path='./reports', filename='report_air_offdesign.tex')

document_model(nw, path='./reports', filename='report_air_offdesign.tex') is flaging
AttributeError: 'ReferencedFluidProperties' object has no attribute 'val'

**Reply by fwitte** [reply]
Hi @Adefunmike and thank you for reaching out. Your request is not really helpful, please:

- state what the context is, e.g. with a short description of what you want to model and a flowsheet of the plant
- write, what the issue actually is (e.g. an error messages etc.)

Otherwise nobody will be able to help you here.

Thank you! Best

Francesco
